_source/_posts/2023-07-28-scim-workshop.md

```diff
@@ -0,0 +1,1150 @@
+
+
+Now we are ready to test with Postman with our local server. You can also make cURL requests directly on the terminal if you prefer. 
+
+##### Try a  POST request 
+Try sending the following request to `http://localhost:3333/scim/v2/Users`:
+
+```
+{
+    "schemas": [
+        "urn:ietf:params:scim:schemas:core:2.0:User"
+    ],
+    "userName": "test.user@portal.example",
+    "name": {
+        "givenName": "Test",
+        "familyName": "User"
+    },
+    "emails": [
+        {
+            "primary": false,
+            "value": "test.user@portal.example",
+            "type": "work"
+        },
+        {
+            "primary": true,
+            "value": "test.user@portal.example",
+            "type": "work"
+        }
+    ],
+    "displayName": "Test User",
+    "locale": "en-US",
+    "externalId": "00ujl29u0le5T6Aj10h7",
+    "groups": [],
+    "password": "1mz050nq",
+    "active": true
+}
+```
+
+What response do you expect? If your code is working correctly, the Todo app backend will give a response with status 201 Created, and the following body: 
+
+```
+{
+    "schemas": [
+        "urn:ietf:params:scim:schemas:core:2.0:User"
+    ],
+    "locale": "en-US",
+    "groups": [],
+    "meta": {
+        "resourceType": "User"
+    },
+    "id": "5",
+    "userName": "test.user@portal.example",
+    "name": {
+        "givenName": "Test",
+        "familyName": "User"
+    },
+    "emails": [
+        {
+            "primary": true,
+            "value": "test.user@portal.example",
+            "type": "work"
+        }
+    ],
+    "displayName": "Test User",
+    "externalId": "00ujl29u0le5T6Aj10h7",
+    "active": true
+}
+```
+
+If you send the same POST request again, re-creating an existing user, what should happen? Try it, and you'll get a 409 Conflict response, with a body like this: 
+
+```
+{
+    "schemas": [
+        "urn:ietf:params:scim:api:messages:2.0:Error"
+    ],
+    "detail": "User already exists in the database: 5",
+    "status": 409
+}
+```
+
+You can repeat this testing process for each SCIM route that you implement! 
+
+### GET /Users
+According to [the SCIM spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.4.2.4), you may return users paginated. [Okta requests users paginated]( https://developer.okta.com/docs/reference/scim/scim-20/#retrieve-users) with a Start Index set to 1 and a count set to 100 per page. 
+
+In addition, filtering by username is an OPTIONAL parameter for SCIM service providers refer to this section of [the SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.2.2). However, Okta requires username filtering support as it checks to see if the user exists by username first before creating them. Ref: [Okta Docs](https://developer.okta.com/docs/reference/scim/scim-20/#create-users)
+
```

Comment:
```suggestion

Why would a client, in this case the identity provider, ever want to ask our app's SCIM server what users it knows about? Okta uses this endpoint to test whether a user exists, so that it can send the correct request type: PUT to update an existing user, or POST to create a new account for a user that wasn't previously known to the SCIM server. 

```